sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionLifeCycleManager.cs

```diff
@@ -0,0 +1,251 @@
+            _sessionId = sessionId;
+            _identifier = identifier;
+            _maxAutoLockRenewal = maxAutoLockRenewal;
+            _sessionInitHandler = sessionInitHandler;
+            _sessionCloseHandler = sessionCloseHandler;
+            _errorHandler = errorHandler;
+        }
+
+        public async Task<ServiceBusReceiver> GetOrCreateSessionReceiver(
+            ServiceBusReceiverOptions receiverOptions,
+            CancellationToken cancellationToken)
+        {
+            if (_sessionId == null)
+            {
+                _receiver = await CreateAndInitializeSessionReceiver(
+                    receiverOptions,
+                    cancellationToken,
+                    _sessionId)
+                    .ConfigureAwait(false);
+                return _receiver;
+            }
+            bool releaseSemaphore = false;
+            try
+            {
+                await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                releaseSemaphore = true;
+                _threadCount++;
+                if (_receiver != null)
+                {
+                    return _receiver;
+                }
+
+                _receiver = await CreateAndInitializeSessionReceiver(
+                    receiverOptions,
+                    cancellationToken,
+                    _sessionId).ConfigureAwait(false);
+
+                return _receiver;
+            }
+            finally
+            {
+                if (releaseSemaphore)
+                {
+                    _semaphore.Release();
+                }
+            }
+        }
+
+        private async Task<ServiceBusSessionReceiver> CreateAndInitializeSessionReceiver(
+            ServiceBusReceiverOptions receiverOptions,
+            CancellationToken cancellationToken,
+            string sessionId = default)
+        {
+            _receiver = await ServiceBusSessionReceiver.CreateSessionReceiverAsync(
+                entityPath: _entityPath,
+                connection: _connection,
+                sessionId: sessionId,
+                options: receiverOptions,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            if (_maxAutoLockRenewal > TimeSpan.Zero)
+            {
+                var sessionLockCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                Task _ = RenewSessionLock(
+                    receiver: _receiver,
+                    sessionLockCancellationSource);
+            }
+
+            if (_sessionInitHandler != null)
+            {
+                var args = new ProcessSessionEventArgs(_receiver, cancellationToken);
+                await _sessionInitHandler(args).ConfigureAwait(false);
+            }
+
+            return _receiver;
+        }
+
+        public async Task CloseSessionIfNeeded(CancellationToken cancellationToken)
+        {
+            bool releaseSemaphore = false;
+            if (_receiver == null)
```

Comment:
This field is used under lock and not under lock in different places.

Reply:
Fixed